File: 1462-course-schedule-iv/1462-course-schedule-iv.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @param {number[][]} queries
 * @return {boolean[]}
 */
var checkIfPrerequisite = function(numCourses, prerequisites, queries) {
    const graph = {}
    for (let i=0; i < numCourses; i++) {
        graph[i] = []
    }
    for (const [u, v] of prerequisites) {
        graph[u].push(v)
    }
    return queries.map(q => dfs(...q))
    
    function dfs(u, v, set=new Set()) {
        if (u === v) return true
        if (set.has(u)) return false
        set.add(u)
        for (const next of graph[u]) {
            if (dfs(next, v, set)) return true
        }
        return false
    }
};

File: 1462-course-schedule-iv/1462-course-schedule-iv.py
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        
        graph = {}
        
        for i in range(numCourses):
            graph[i] = []
            
        for u, v in prerequisites:
            graph[u].append(v)
            
        def dfs(u: int, v: int, set_: set) -> bool:
            if u == v:
                return True
        
            if u in set_:
                return False
            
            set_.add(u)
            
            for next_ in graph[u]:
                if dfs(next_, v, set_):
                    return True
            
            return False
        
        return [dfs(u, v, set()) for u, v in queries]